lib/Surface/USB/Shuttle.js

```diff
@@ -0,0 +1,221 @@
+	totalCols: 4,
+	totalRows: 2,
+
+	// TODO(Someone with hardware): This mapping is guesswork and hasn't been tested
+	encoders: [5, 6],
+	buttons: [4, 0, 1, 2, 7],
+}
+const shuttleProV1Info = {
+	// Same as Pro V2 only without the buttons either side of the encoders
+	// Map the same for consistency and compatibility
+	totalCols: 5,
+	totalRows: 4,
+
+	// TODO(Someone with hardware): This mapping is guesswork and hasn't been tested
+	encoders: [11, 12],
+	buttons: [0, 1, 2, 3, 5, 6, 7, 8, 9, 15, 18, 16, 17],
+}
+const shuttleProV2Info = {
+	// 4 buttons
+	// 5 buttons
+	// button, two encoders (jog and shuttle), button
+	// 2 buttons (combine with the row below)
+	// 2 buttons
+	totalCols: 5,
+	totalRows: 4,
+
+	encoders: [11, 12],
+	buttons: [0, 1, 2, 3, 5, 6, 7, 8, 9, 15, 18, 16, 17, 10, 13],
+}
+
+const encoders = Object.freeze({
+	JOG: 1,
+	SHUTTLE: 2,
+})
+
+function buttonToIndex(modelInfo, info) {
+	const index = modelInfo.buttons[info - 1]
+	if (index !== undefined) {
+		return index
+	}
+
+	return undefined
+}
+
+function rotaryToButtonIndex(modelInfo, info) {
+	const index = modelInfo.encoders[info - 1]
+	if (index !== undefined) {
+		return index
+	}
+
+	return undefined
+}
+
+class SurfaceUSBShuttle {
+	constructor(ipcWrapper, devicepath, shuttle, modelInfo, deviceInfo) {
+		this.ipcWrapper = ipcWrapper
+		this.shuttle = shuttle
+		this.deviceInfo = deviceInfo
+		this.modelInfo = modelInfo
+
+		this.config = {
+			// No config currently present
+		}
+
+		this.ipcWrapper.log('debug', `Adding Contour Shuttle USB device ${devicepath}`)
+
+		this.info = {
+			type: `Shuttle ${this.deviceInfo.name}`,
+			devicepath: devicepath,
+			configFields: [],
+			keysPerRow: this.modelInfo.totalCols,
+			keysTotal: this.modelInfo.totalCols * this.modelInfo.totalRows,
+			deviceId: `shuttle:${this.deviceInfo.id}`,
+		}
+
+		this.shuttle.on('error', (error) => {
+			console.error(error)
+			this.ipcWrapper.remove()
+		})
+
+		this.shuttle.on('buttondown', (info) => {
+			const keyIndex = buttonToIndex(this.modelInfo, info) ?? rotaryToButtonIndex(this.modelInfo, info)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.click(keyIndex, true)
+		})
+
+		this.shuttle.on('buttonup', (info) => {
+			const keyIndex = buttonToIndex(this.modelInfo, info) ?? rotaryToButtonIndex(this.modelInfo, info)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.click(keyIndex, false)
+		})
+
+		this.shuttle.on('jog-dir', (delta) => {
+			const keyIndex = rotaryToButtonIndex(this.modelInfo, encoders.JOG)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.rotate(keyIndex, delta == 1)
+		})
+
+		this.shuttle.on('shuttle-trans', (previous, current) => {
+			const keyIndex = rotaryToButtonIndex(this.modelInfo, encoders.SHUTTLE)
+			if (keyIndex === undefined) {
+				return
+			}
+
+			this.ipcWrapper.rotate(keyIndex, previous < current)
```

Comment:
Im not sure about the strategy here, but maybe this will do for now?

Reply:
I don't think I necessarily found a better explanation of what I should be doing with the rotate IPC otherwise?